### PR TITLE
First pass over Refining.

### PR DESCRIPTION
**Pull request by nicelhc13**
First refactoring Refining codes. I mainly removed duplicated or unnecessary codes, applied snake case, and other code conventions.

### CODE REVIEW

File: lonestar/eda/cpu/bipart-nested/Bipart.h [hunk 1]
```diff
@@ -20,8 +20,35 @@
 #ifndef BIPART_H_
 #define BIPART_H_
 
+#include <algorithm>
+#include <array>
+#include <climits>
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Bipart.h:25]: Remove if any of these headers are unnecessary which I am sure is the case.

File: lonestar/eda/cpu/bipart-nested/Bipart.h [hunk 2]
```diff
@@ -20,8 +20,35 @@
...
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <numeric>
+#include <map>
+#include <set>
+#include <stack>
+#include <stdlib.h>
+#include <string.h>
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Bipart.h:35]: #include <string>

File: lonestar/eda/cpu/bipart-nested/Bipart.h [hunk 3]
```diff
@@ -20,8 +20,35 @@
...
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "galois/Galois.h"
+#include "galois/Reduction.h"
+#include "galois/graphs/ReadGraph.h"
+#include "galois/graphs/FileGraph.h"
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Bipart.h:43]: Is this required?

File: lonestar/eda/cpu/bipart-nested/Bipart.h [hunk 4]
```diff
@@ -20,8 +20,35 @@
...
 #include "galois/graphs/LC_CSR_Graph.h"
 #include "galois/AtomicWrapper.h"
+#include "galois/AtomicHelpers.h"
+#include "galois/Timer.h"
+#include "galois/DynamicBitset.h"
+#include "galois/runtime/Profile.h"
+#include "galois/substrate/PerThreadStorage.h"
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Bipart.h:50]: She other Lonestar apps and what boilerplate headers that they include and I am sure we can remove almost everything here.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 5]
```diff
@@ -42,3 +42,128 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+  std::sort(nodes.begin(), nodes.end(), [&graph](GNode& l_opr, GNode& r_opr) {
+    MetisNode& l_data               = graph.getData(l_opr);
+    MetisNode& r_data               = graph.getData(r_opr);
+    float l_gain                    = l_data.getGain();
+    float r_gain                    = r_data.getGain();
+    float l_weight                  = l_data.getWeight();
+    float r_weight                  = r_data.getWeight();
+    float l_cost                    = l_gain / l_weight;
+    float r_cost                    = r_gain / r_weight;
+    [[maybe_unused]] uint32_t l_nid = l_data.nodeid;
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:56]: remove these if not used.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 6]
```diff
@@ -42,3 +42,128 @@ void MRI_f(GNode node, GGraph* fine_graph) {
...
+      return l_nid < r_nid;
+    }
+
+    return l_cost > r_cost;
+  });
+}
+
+void InitGain(GGraph& g) {
+  galois::do_all(
+      galois::iterate(g),
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:69]: Start iterator from g.hedges instead of return in case n < g.hedges.

Comment (gurbinder533) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:69]: galois::iterate(g.hedges, (g.size()) ...

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 7]
```diff
@@ -17,717 +17,428 @@
...
-#include "galois/Reduction.h"
-#include "galois/Timer.h"
 #include "Bipart.h"
-#include "galois/AtomicHelpers.h"
-
-namespace {
+#include "Helper.h"
 
-constexpr static const unsigned CHUNK_SIZE = 256U;
+constexpr static const uint32_t CHUNK_SIZE = 256U;
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:23]: kChunkSize and U--> u

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 8]
```diff
@@ -17,717 +17,428 @@
...
     std::vector<std::pair<uint32_t, uint32_t>>& combined_edgelist,
     std::vector<std::pair<uint32_t, uint32_t>>& combined_nodelist,
-    std::vector<GGraph*>& g, unsigned refineTo) {
-  typedef galois::gstl::Vector<unsigned> VecTy;
-  typedef galois::substrate::PerThreadStorage<VecTy> ThreadLocalData;
-  ThreadLocalData edgesThreadLocal;
-  std::string name = "findZandO";
-
-  // typedef galois::worklists::PerSocketChunkFIFO<8> Chunk;
+    std::vector<GGraph*>& g, uint32_t refineTo) {
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:50]: better name of refineTo?

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 9]
```diff
@@ -17,717 +17,428 @@
...
+        GNode n               = node_index_pair.first;
+        MetisNode& node_data  = g[index]->getData(n);
+        uint32_t weight       = node_data.getWeight();
+
+        node_size[index] += weight;
+        if (node_data.getPart() > 0) {
+          accum[index] += weight;
+        }
       },
       galois::steal(), galois::loopname("make balance"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:71]: make balance --> Make-balance

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 10]
```diff
@@ -17,717 +17,428 @@
...
-  unsigned pass = 0;
-  // initGain(g);
-  while (pass < refineTo) {
 
-    initGains(combined_edgelist, combined_nodelist, g);
+  GNodeBag part_zero_nodes;
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:73]: change part to partition everywhere, please. Makes it easy to read.

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 11]
```diff
@@ -17,717 +17,428 @@
...
+  uint32_t num_partzero_nodes{0};
+  uint32_t num_partone_nodes{0};
 
-    // rest of the code is very fast
-    for (int i = 0; i < len; i++) {
+  for (uint32_t pass = 0; pass < refineTo; pass++) {
+    InitGain(combined_edgelist, combined_nodelist, g);
 
-      if (g[i] == NULL)
+    // rest of the code is very fast
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:84]: Remove this comment.

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 12]
```diff
@@ -17,717 +17,428 @@
...
+      }
+      galois::do_all(
+          galois::iterate(swap_bag),
+          [&](GNode n) {
+            MetisNode& node_data = cur_graph.getData(n);
+            uint32_t part        = node_data.getPart();
+            node_data.setPart(1 - part);
+            node_data.counter++;
+          },
+          galois::steal(), galois::loopname("swap"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:166]: Swap

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 13]
```diff
@@ -17,717 +17,428 @@
...
+          } else if (gain > -9.0f) {
+            int32_t d    = gain * 10.0f - 1;
+            uint32_t idx = 10 - d;
+            cand_nodes_bag_arr[idx].emplace(n);
+          } else { // NODES with gain by weight ratio <= -9.0f are in one
+                   // bucket
+            neg_cand_nodes_bag.emplace(n);
+          }
+        },
+        galois::steal(), galois::loopname("bucket gain"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:261]: Bucket-Gain

File: lonestar/eda/cpu/bipart-nested/Refine.cpp [hunk 14]
```diff
@@ -17,717 +17,428 @@
...
-              int d = gain * 10.0f;
-              idx   = 10 - d;
+          GNode n              = *cand_nodes_bag.begin();
+          MetisNode& node_data = g.getData(n);
+          float gain =
+              static_cast<float>(node_data.getGain()) / node_data.getWeight();
+          uint32_t idx{0};
+          if (gain < 1.0f) {
+            if (gain >= 0.0f) {
+              int32_t d = gain * 10.0f;
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Refine.cpp:278]: better variable name for "d"?

File: lonestar/eda/cpu/bipart-nested/Bipart.h (hunk 2 above)
Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Bipart.h:35]: Sorry, what do you mean?

File: lonestar/eda/cpu/bipart-nested/Bipart.h (hunk 3 above)
Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Bipart.h:43]: The first one is necessary. Below three are not.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp (hunk 5 above)
Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:56]: It is conditionally used. So we cannot remove it.

File: lonestar/eda/cpu/bipart-nested/Refine.cpp (hunk 8 above)
Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Refine.cpp:50]: @uditagar  any idea?

File: lonestar/eda/cpu/bipart-nested/Refine.cpp (hunk 14 above)
Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Refine.cpp:278]: @uditagar

File: lonestar/eda/cpu/bipart-nested/Bipart.h (hunk 2 above)
Comment (gurbinder533) [reply lonestar/eda/cpu/bipart-nested/Bipart.h]: include string instead of string.h

File: lonestar/eda/cpu/bipart-nested/Refine.cpp (hunk 8 above)
Comment (gurbinder533) [reply lonestar/eda/cpu/bipart-nested/Refine.cpp:48]: refine_level?

File: lonestar/eda/cpu/bipart-nested/Refine.cpp (hunk 14 above)
Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Refine.cpp:277]: gain_10X?

File: lonestar/eda/cpu/bipart-nested/Refine.cpp (hunk 8 above)
Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Refine.cpp:48]: refine_max_levels. (same as we did for coarsening)

**Pull request merged**